### PR TITLE
DEV2-4144 fix sign in flow

### PR DESCRIPTION
**Pull request by amirbilu**

### CODE REVIEW

**Comment by amirbilu** [comment]
image

File: Common/src/main/java/com/tabnineCommon/chat/ChatFrame.kt [hunk 1]
```diff
@@ -59,10 +72,10 @@ class ChatFrame(private val project: Project, private val binaryRequestFacade: B
     }
 
     private fun updateDisplay() {
-        if (ChatEnabled.getInstance().enabled) {
+        if (ChatEnabled.getInstance().enabled && isLoggedIn) {
             displayChat()
         } else {
-            if (capabilitiesFetched || Config.IS_SELF_HOSTED) {
+            if (capabilitiesFetched || Config.IS_SELF_HOSTED || !isLoggedIn) {
```
Comment (amirbilu) [inline Common/src/main/java/com/tabnineCommon/chat/ChatFrame.kt:78]: Asking myself what's the meaning of `Config.IS_SELF_HOSTED` here. I'm not sure.

File: Common/src/main/java/com/tabnineCommon/chat/ChatDisabledComponents.kt [hunk 2]
```diff
@@ -1,60 +1,44 @@
...
     }
+    pane.addHyperlinkListener(
+        HyperlinkListener { it ->
+            if (it.eventType !== HyperlinkEvent.EventType.ACTIVATED) {
+                return@HyperlinkListener
+            }
+            if (it.url.toString() == "https://www.tabnine.com/#ChatSection") {
+                BrowserUtil.browse(it.url.toString())
+            } else if (it.url.toString() == "https://app.tabnine.com/login") {
+                DependencyContainer.instanceOfBinaryRequestFacade().executeRequest(LoginRequest {})
```
Comment (yanir-codota) [inline Common/src/main/java/com/tabnineCommon/chat/ChatDisabledComponents.kt:38]: is this the same as the login from the status bar?

File: Common/src/main/java/com/tabnineCommon/chat/ChatFrame.kt [hunk 3]
```diff
@@ -27,13 +28,25 @@ import javax.swing.event.HyperlinkEvent
...
 
         updateDisplay()
 
         val connection = ApplicationManager.getApplication().messageBus.connect(this)
+
+        connection.subscribe(
+            BinaryStateChangeNotifier.STATE_CHANGED_TOPIC,
+            BinaryStateChangeNotifier { state ->
+                isLoggedIn = state.isLoggedIn == true
+                ApplicationManager.getApplication().invokeLater {
```
Comment (yanir-codota) [inline Common/src/main/java/com/tabnineCommon/chat/ChatFrame.kt:44]: we're having serious issues with the locks on the UI thread. 
is this a safe way to do it

**Review by yanir-codota — COMMENTED** [review]
Looks good! 
Had some questions, just to make sure we're covered.

File: Common/src/main/java/com/tabnineCommon/chat/ChatDisabledComponents.kt (hunk 2 above)
Comment (amirbilu) [reply Common/src/main/java/com/tabnineCommon/chat/ChatDisabledComponents.kt:38]: yes!

File: Common/src/main/java/com/tabnineCommon/chat/ChatFrame.kt (hunk 3 above)
Comment (amirbilu) [reply Common/src/main/java/com/tabnineCommon/chat/ChatFrame.kt:44]: yes!

**Review by yanir-codota — APPROVED** [review]
LGTM

**Pull request merged**